### PR TITLE
Fixed #24349 -- Limited domain name labels to 63 characters in EmailValidator

### PR DESCRIPTION
**Pull request by atmb4u**
@timgraham Ref: [pull request](https://github.com/django/django/pull/4273)

### CODE REVIEW

File: tests/validators/tests.py [hunk 1]
```diff
@@ -45,7 +45,10 @@
     (validate_email, 'email@localhost', None),
     (EmailValidator(whitelist=['localdomain']), 'email@localdomain', None),
     (validate_email, '"test@test"@example.com', None),
+    (validate_email, 'example@atm.longgtldfortesting', None),
```
Comment (timgraham) [inline tests/validators/tests.py]: I'd suggest something like `'example@atm.%s' % 'a' * 63` (and 64 for the invalid one)

File: docs/releases/1.9.txt [hunk 2]
```diff
@@ -236,6 +236,8 @@ Validators
 * Added :func:`django.core.validators.int_list_validator` to generate
   validators of strings containing integers separated with a custom character.
 
+* Limits email domain TLD length to 63.
```
Comment (timgraham) [inline docs/releases/1.9.txt]: ``EmailValidator`` now limits... ... 63 characters.

File: django/core/validators.py [hunk 3]
```diff
@@ -144,7 +144,9 @@ class EmailValidator(object):
     domain_regex = re.compile(
         # max length of the domain is 249: 254 (max email length) minus one
         # period, two characters for the TLD, @ sign, & one character before @.
-        r'(?:[A-Z0-9](?:[A-Z0-9-]{0,247}[A-Z0-9])?\.)+(?:[A-Z]{2,6}|[A-Z0-9-]{2,}(?<!-))$',
+        # max length of the gTDs is 63
```
Comment (timgraham) [inline django/core/validators.py]: is this limited to TLDs? it seems all labels are limited to 63 characters.

Comment (atmb4u) [inline django/core/validators.py]:
```python
r'((?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.)+)(?:[A-Z0-9-]{2,63}(?<!-))$'
```
This is what I came up with. Each of the label can be as long as 63. But now, total length can be greater than 255.
Any suggestions on how to handle it from regex? Or should I check the length just before returning True or False in line no. 188?

Comment (timgraham) [inline django/core/validators.py]: Checking the overall length outside the regex seems okay. Where does 61 in the proposed regex come from?

Comment (atmb4u) [inline django/core/validators.py]: 1(matched in the first  group - can't begin with ```-``` ) +61+1(matched in the last  group -can't end with```-```) = **63**

File: django/core/validators.py [hunk 4]
```diff
@@ -142,9 +142,9 @@ class EmailValidator(object):
         r'|^"([\001-\010\013\014\016-\037!#-\[\]-\177]|\\[\001-\011\013\014\016-\177])*"$)',  # quoted-string
         re.IGNORECASE)
     domain_regex = re.compile(
-        # max length of the domain is 249: 254 (max email length) minus one
-        # period, two characters for the TLD, @ sign, & one character before @.
-        r'(?:[A-Z0-9](?:[A-Z0-9-]{0,247}[A-Z0-9])?\.)+(?:[A-Z]{2,6}|[A-Z0-9-]{2,}(?<!-))$',
+        # max length for DNS labels and TLD is 63 octets
```
Comment (timgraham) [inline django/core/validators.py]: octets -> characters?

File: docs/releases/1.9.txt [hunk 5]
```diff
@@ -236,6 +236,8 @@ Validators
 * Added :func:`django.core.validators.int_list_validator` to generate
   validators of strings containing integers separated with a custom character.
 
+* EmailValidator now limits DNS labels and TLD length to 63 octets.
```
Comment (timgraham) [inline docs/releases/1.9.txt]: double backticks around EmailValidator since it's a class name.

File: tests/validators/tests.py [hunk 6]
```diff
@@ -45,7 +45,12 @@
     (validate_email, 'email@localhost', None),
     (EmailValidator(whitelist=['localdomain']), 'email@localdomain', None),
     (validate_email, '"test@test"@example.com', None),
+    (validate_email, 'example@atm.%s' % ('a' * 63), None),
+    (validate_email, 'example@%s.atm' % ('a'*63), None),
```
Comment (timgraham) [inline tests/validators/tests.py]: missing spaces around *

**Comment by timgraham** [comment]
buildbot, add to whitelist.

**Comment by timgraham** [comment]
It seems like we should limit the email addresses to 254 characters as well, but that's a separate ticket.

File: docs/releases/1.9.txt [hunk 7]
```diff
@@ -236,6 +236,9 @@ Validators
 * Added :func:`django.core.validators.int_list_validator` to generate
   validators of strings containing integers separated with a custom character.
 
+* :class:`validators.EmailValidator <django.core.validators.EmailValidator>` now
```
Comment (timgraham) [inline docs/releases/1.9.txt]:
```
:class:`~django.core.validators.EmailValidator`
```
is fine.

**Comment by timgraham** [comment]
Could you please squash your commits? Thanks!

**Comment by timgraham** [comment]
merged in 24003295088ec24d823f66bd5c8b917478b24c5d, thanks!

**Pull request closed without merging**